classification/config.py

```diff
@@ -3,116 +3,188 @@
+        if  self.config_dict[parameter] is None:
+            raise ValueError(f'The required parameter "{parameter}" is not defined in yaml')
+
+    def generate_config_file(self, filename="test.yml"):
+        """
+        Sends all configs saved to class to new file
+        overwrites file
+        """
+        with open(filename, 'w', encoding='utf-8') as file:
+            yaml.dump(self.config_dict, file)
     
-    #Add git hash to config so wand logging can track vrs used for reproduciblity
-    try:
-        repo = git.Repo(search_parent_directories=True)
-        sha = repo.head.object.hexsha
-        setattr(CONFIG, "git_hash", sha)
-        print(sha)
-
-    #I want to catch this spefific error, and it doesn't extend from base exception
-    #¯\(ツ)/¯
-    # pylint: disable=no-member
-    except git.exc.InvalidGitRepositoryError:
-        print("InvalidGitRepositoryError found, this means we cannot save git hash :(")
-        print("You are likely calling a python file outside of this repo") 
-        print("if from command line, cd into acoustic-mutliclass-training")
-        print("then you can run the script again")
-        sys.exit(1)
-
-    return CONFIG
+    def __getattr__(self,attr):
+        """
+        Gets a config value in a dict-like way
+
+        attr string for varible name as defined in a config.yml file
+        """
+        return self.config_dict[attr]
+
+    def get_git_hash(self):
+        """
+        Gets the hash of the current git commit
+        This requires recording the hash before model training
+        """
+
+        #Add git hash to config so wand logging can track vrs used for reproduciblity
+        try:
+            repo = git.Repo(search_parent_directories=True)
+            sha = repo.head.object.hexsha
+            self.config_dict["git_hash"] = sha
+
+        #I want to catch this spefific error, and it doesn't extend from base exception
+        #¯\(ツ)/¯
+        # pylint: disable=no-member
+        except git.exc.InvalidGitRepositoryError:
+            print("InvalidGitRepositoryError found, this means we cannot save git hash :(")
+            print("You are likely calling a python file outside of this repo") 
+            print("if from command line, cd into acoustic-mutliclass-training")
+            print("then you can run the script again")
+            sys.exit(1)
+        return sha
+
+def get_config():
+    """ Returns a config variable with the command line arguments or defaults
+    Decrepated, returns Config to prevent largescale code breaks
+    """
+    return cfg
+
+def testing():
+    """
+    Test functionality of generating and caching configs
+    """
+
+    config = Config()
+
+    # I want to test singleton class creation
+    # So I want to change one instance var and see that var in a new class
+    # in practice, I never want to change a config setting outside of config
+    # pylint: disable=attribute-defined-outside-init
+    config.change = " hah"
```

Comment:
what does "hah" mean?

Reply:
hah

Reply:
Honestly I just had it in there like "lol" to test the code

Reply:
Goal of that is to test if the singleton class is working